invenio_rdm_records/services/components/pids.py

```diff
@@ -9,177 +9,257 @@
+        """This method deletes PIDs of a draft.
 
-            is_required = provider_config["required"]
-            is_system_managed = provider_config["system_managed"]
+        It should only delete pids with status `NEW`, other pids would
+        belong to previous versions of the record.
+        """
+        pids = draft.get('pids', {})
+        for scheme, pid_attrs in pids.items():
+            provider_name = pid_attrs.get("provider")
+            provider = self.service.pids.get_provider(scheme, provider_name)
+            try:
+                pid = provider.get(pid_value=pid_attrs["identifier"])
+                # FIXME: should check returned value and log ig deletion failed
+                # pids should be status NEW at this point
+                if pid.is_new():
+                    provider.delete(pid, record=draft)
+            except PIDDoesNotExistError:
+                pass  # pid was not saved to pidstore yet, no deletion needed
 
-            if is_system_managed:
-                self._publish_managed(scheme, provider, is_required, draft_pid,
-                                      record_pids, draft=draft)
-            else:
-                self._publish_unmanaged(scheme, provider, is_required,
-                                        draft_pid, record_pids, draft=draft)
+    def _remove_if_created(
+        self, pid_value, pid_type, pid_provider, record=None
+    ):
+        """Deletes a PID if it has been created/reserved."""
+        if pid_value:  # e.g. ask to remove an external and incomplete pid
+            provider = self.service.pids.get_provider(pid_type, pid_provider)
+            try:
+                pid = provider.get(pid_value=pid_value)
+                provider.delete(pid, record=record)
+            except PIDDoesNotExistError:
+                pass  # if it does not exist, it does not need to be deleted
 
-        record.pids = record_pids
+    def update_draft(self, identity, data=None, record=None, errors=None):
+        """Update draft handler.
+
+        The permission check is performed on pid updates, no new additions.
+        In addition, they check only if it is allowed to remove the pid
+        since adding a new one has the same restrictions than the record
+        update.
+        """
+        # loop through the record (draft) pids and see if they need updating
+        # dict() forces copy to avoid modification during loop errors
+        existing_pids = dict(record.get("pids", {}))
+        data_pids = data.get("pids", {})
+        errors = errors or []
+
+        # validate new pids
+        self._validate_pids_schemes(data_pids)
+        self._validate_pids(data_pids, record, errors)
+
+        # update content
+        for scheme, pid in existing_pids.items():
+            updated_pid = data_pids.get(scheme)
+            if updated_pid != pid:
+                try:  # record and scheme reach the generator as "over"
+                    self.service.require_permission(
```

Comment:
I haven't checked. But...
- Since the generators are not needed in the end, and it is checked at *provider* level there are not many queries done (just one per PID, to get it and check status).
- Per service method, only one perm check is done (as in any other service method).